File: solutions/javascript/binary-search/1/binary-search.js
export const find = (array, value) => {
  if (!array || array.length === 0) {
    throw new Error('Value not in array');
  }

  let left = 0;
  let right = array.length - 1;

  while (left <= right) {
    const mid = Math.floor((left + right) / 2);
    const midValue = array[mid];

    if (midValue === value) {
      return mid; // Found the value, return its index
    } else if (midValue < value) {
      // Value is in the right half
      left = mid + 1;
    } else {
      // Value is in the left half
      right = mid - 1;
    }
  }

  throw new Error('Value not in array');
};

File: solutions/java/binary-search/1/src/main/java/BinarySearch.java
import java.util.List;

class BinarySearch {
    private final List<Integer> items;

    BinarySearch(List<Integer> items) {
        this.items = items;
    }

    int indexOf(int item) throws ValueNotFoundException {
        int left = 0;
        int right = items.size() - 1;

        while (left <= right) {
            // Prevent integer overflow for large lists
            int mid = left + (right - left) / 2;
            int midValue = items.get(mid);

            if (midValue == item) {
                return mid;
            } else if (midValue < item) {
                // Item is in the right half
                left = mid + 1;
            } else {
                // Item is in the left half
                right = mid - 1;
            }
        }

        throw new ValueNotFoundException("Value not in array");
    }
}